platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+
+Events are not only given to the Event Intake system through gossip. Internal events are also fed to Event Intake.
+These internal events **may** bypass some steps in the pipeline. For example, internal self-events (those events created
+by the node itself) do not need validation. Likewise, when replaying events from the pre-consensus event buffer, those
+checks are not needed (since they have already been proved valid and are in topological order).
+
+#### Peer Discipline
+
+During the validation process, if an event is invalid, it is rejected, and this information is passed to the Sheriff
+module so the offending node may be disciplined. Note that the node to be disciplined will be the node that sent this
+bad event to us, not the origin node. This information (which node sent the event) must be captured by Gossip and
+passed to Event Intake as part of the event metadata.
+
+#### Topological Ordering
+
+Events are buffered if necessary to ensure that each parent event has been emitted from Event Intake before any child
+events. A simple map (the same used for deduplication) can be used here. Given some event, for each parent, look up the
+parent by its hash. If each parent is found in the map, then emit the event. Otherwise, remember the event so when the
+missing parent is received, the child may be emitted. The current implementation uses what is known as the "orphan
+buffer" for this purpose.
+
+Since Event Intake will also maintain some buffers, it needs to know about the progression of the hashgraph,
+so it can evict old events. In this case, the "orphan buffer" holds events until either the parent events have
+arrived, or the events expired due to the advancement of the "non-ancient event window" and the event is
+dropped from the buffer. This document does not prescribe the existence of the orphan buffer or the method by which
+events are sorted and emitted in topological order, but it does describe a method by which old events can be dropped.
+
+#### Assigning Generations
+
+Several algorithms, including the hashgraph consensus algorithm and the tipset algorithm, require events to be part of
+a "generation". We use what is known as "local generation". Each node, when it starts, assigns "1" as the generation
+of the first event it encounters per event creator. So the first event from Alice is generation 1, as is the first event
+from Bob, but the second event from Alice is in generation 2. This resets each time the consensus module is restarted
+(node restart, upgrade, reconnect, etc.). The absolute numbers for the generations are not important, but their
+relative values are critical. Generations always increase by 1. Different nodes may assign different generations to
+the same events. The generation will be part of the event metadata, **not** part of `EventCore` (that is, it is not
+in the part of the event that is gossiped).
+
+Event Intake is responsible for assigning generations to events.
+
+#### Fork Detection
```

Comment:
Although we historically called these "forks", it was renamed to "branching" in order to avoid confusion with the more commonly understood concept of blockchain forking.